### ISSUE TITLE
Vertical glitch in terminal (insiders / Windows)

### ISSUE THREAD

**Issue by warpdesign** [Microsoft/vscode#34639]
- VSCode Version: Insiders 1.17.0 (commit 055825...)
- OS Version: Windows 10 (64bit)

Steps to Reproduce:

1. Open integrated terminal
2. Type `dir`

What happens?
There is a vertical glitch that appears after some characters:

![vscode](https://user-images.githubusercontent.com/199648/30598020-fe58367a-9d58-11e7-996b-bc753b6fb3be.png)

**Comment by warpdesign** [comment]
It's supposed to be a space, see the comparison with the cmd prompt output:

![vscodeglitch](https://user-images.githubusercontent.com/199648/31006260-484d7f24-a4fc-11e7-9aec-ced12454fdfe.png)

**Comment by Tyriar** [comment]
Hmm, it's hard to say whether this is fixed already. I can't reproduce and quite a lot of fixes have gone into Insiders since the build was blocked due to signing problems.

I might close this off so that people on Windows machines can verify it as I believe it's fixed. When you get the new build you can test again and we can reopen if necessary.

**Comment by warpdesign** [comment]
@Tyriar Where can I get an up to date build ? I'll can tell you if this still happens.

**Comment by Tyriar** [comment]
Try this one in the zip: https://az764295.vo.msecnd.net/insider/3736ccfe9731b479040890d745b0d4f4173fc040/VSCode-win32-x64-1.17.0-insider-unsigned.zip

Auto updated signed builds will hopefully be available again soon.

**Comment by warpdesign** [comment]
The bug is still present in this build:

![vscodeglithc](https://user-images.githubusercontent.com/199648/31012156-1436607e-a511-11e7-896b-88a87ac5cdcd.png)

**Comment by Tyriar** [comment]
@warpdesign hmm... What's puzzling to me is that it's only happening in that particular cell. If this is indeed just a regular space character, it should behave the same as all the other space characters. But it's just the ones separating numbers.

Zooming in you can see that the haracters are correctly constrained to their 7px wide cells:

<img width="523" alt="screen shot 2017-09-29 at 7 41 32 am" src="https://user-images.githubusercontent.com/2193314/31014583-a51564be-a4e9-11e7-83da-a69b2cb6c993.png">

I won't have access to a Windows machine for a couple of days, but maybe you could help diagnose the issue. Could you open up this file within the install: `app/node_modules/xterm/lib/renderer/BaseRenderLayer.js`, find the drawChar function and add this to the end of it:

```js
this._ctx.clearRect(0, 0, this._canvas.width, this._canvas.height);
this._ctx.drawImage(this._charAtlas, 0, 0);
```

Basically we're trying to add the following to display the bitmap that the chars should be drawn from:

https://github.com/sourcelair/xterm.js/blob/313d1e6fecd98a4a5e8dddebb832a6c4c16a828d/src/renderer/BaseRenderLayer.ts#L269-L271

After that's done, the terminal should show rows of ascii chars drawn in the terminal area. Could you take a screenshot of that while the terminal is fairly wide?

---

Another thing that might help is copying the text and pasting it into a comment inside a code block here. We may be able to verify whether it's a regular space char or not from that.

**Comment by warpdesign** [comment]
Here is the charAtlas:

![canvas-vscode](https://user-images.githubusercontent.com/199648/31015426-5181f6d2-a520-11e7-9381-ac3d6bed759f.png)

Here is the text:

```
14 Dir(s)  52 098 826 240 bytes free
```

**Comment by warpdesign** [comment]
Looking at it in hexa it looks like a normal space (`0x20`).

**Comment by roblourens** [comment]
Reopening, I don't know if you want to put it on October

**Comment by Tyriar** [comment]
@warpdesign thanks, that reveals exactly what the problem is 👍. I was assuming 1px padding on the char cell would be enough to capture anything for monospace fonts. I guess I need to actually clip the cell to cover cases like this. What font is this exactly?

You can work around this by using a font that doesn't use non-monospace boxes for the early ascii chars.

**Comment by warpdesign** [comment]
I don't know what font is being used: I didn't change the default settings which is "" in the settings file :)
And since it's now rendered using canvas I cannot simply inspect the page to see which font is used.

What are the rules for the default fontFamily of the terminal?

**Comment by warpdesign** [comment]
I confirm the problem is gone by changing font family.

**Comment by Tyriar** [comment]
Hmm, I can't repro using Consolas, do you have other font-related settings?

**Comment by roblourens** [comment]
I don't repro this now, but @Tyriar it doesn't look like there has been a change since the last time I reopened it.

**Comment by Tyriar** [comment]
@roblourens the whole line is being redrawn right now so this is definitely fixed for the time being.
